src/core/thread/link_metrics.cpp

```diff
@@ -0,0 +1,299 @@
+{
+    mLinkMetricsReportCallback        = aCallback;
+    mLinkMetricsReportCallbackContext = aCallbackContext;
+}
+
+otError LinkMetrics::SendLinkMetricsQuery(const Ip6::Address &     aDestination,
+                                          uint8_t                  aSeriesId,
+                                          const LinkMetricsTypeId *aTypeIdFlags,
+                                          uint8_t                  aTypeIdFlagsCount)
+{
+    otError                 error = OT_ERROR_NONE;
+    Tlv                     tlv;
+    LinkMetricsQueryId      linkMetricsQueryId;
+    LinkMetricsQueryOptions linkMetricsQueryOptions;
+    uint8_t                 length = 0;
+    static const uint8_t    tlvs[] = {Mle::Tlv::kLinkMetricsReport};
+    uint8_t                 buf[sizeof(Tlv) * 3 + sizeof(LinkMetricsQueryId) + sizeof(LinkMetricsQueryOptions)];
+
+    // Link Metrics Query TLV
+    tlv.SetType(Mle::Tlv::kLinkMetricsQuery);
+    length += sizeof(Tlv);
+
+    // Link Metrics Query Id sub-TLV
+    linkMetricsQueryId.Init();
+    linkMetricsQueryId.SetSeriesId(aSeriesId);
+    memcpy(buf + length, &linkMetricsQueryId, sizeof(Tlv) + linkMetricsQueryId.GetLength());
+    length += sizeof(Tlv) + linkMetricsQueryId.GetLength();
+
+    // Link Metrics Query Options sub-TLV
+    if (aTypeIdFlagsCount > 0)
+    {
+        linkMetricsQueryOptions.Init();
+        linkMetricsQueryOptions.SetLinkMetricsTypeIdList(aTypeIdFlags, aTypeIdFlagsCount);
+
+        memcpy(buf + length, &linkMetricsQueryOptions, sizeof(Tlv) + linkMetricsQueryOptions.GetLength());
+        length += sizeof(Tlv) + linkMetricsQueryOptions.GetLength();
+    }
+
+    // Set Length for Link Metrics Report Tlv
+    tlv.SetLength(length - sizeof(Tlv));
+    memcpy(buf, &tlv, sizeof(Tlv));
+
+    SuccessOrExit(error = Get<Mle::MleRouter>().SendDataRequest(aDestination, tlvs, sizeof(tlvs), 0, buf, length));
+
+exit:
+    return error;
+}
+
+otError LinkMetrics::AppendSingleProbeLinkMetricsReport(Message &                      aMessage,
+                                                        uint8_t &                      aLength,
+                                                        const LinkMetricsQueryOptions *aQueryOptions,
+                                                        const int8_t                   aNoiseFloor,
+                                                        const Message &                aRequestMessage)
+{
+    otError                  error = OT_ERROR_NONE;
+    LinkMetricsReportSubTlv  metric;
+    const LinkMetricsTypeId *linkMetricsTypeIds;
+    uint8_t                  metricsCount = 0;
+
+    OT_ASSERT(aQueryOptions != nullptr);
+    metricsCount       = kLinkMetricsMaxTypeIdFlags;
+    linkMetricsTypeIds = aQueryOptions->GetLinkMetricsTypeIdList(metricsCount);
+
+    for (uint8_t i = 0; i < metricsCount; i++)
+    {
+        // Link Metrics Report sub-TLVs
+        metric.Init();
+        metric.SetMetricsTypeId(linkMetricsTypeIds[i]);
```

Comment:
1.2 spec 4.11.3.4.3.1
> Type / Average Enum
Value 0: Count/summation
Value 1: Exponential Moving Average, initialized with the value of the first sample followed with a weighting of 1/8 for each new sample.

I fails to see how the `Type/Average enum` works -
From the implementation here, it seems nothing special to process this `Type/Average enum`, if so what's the point to have this field?

What's the intention of this field in spec? @jwhui @Irving-cl

Reply:
@librasungirl 

From my point of view, I also don't understand why it's designed in this way. `Type/Average Enum` and `Metric Enum` doesn't work in a combination way. For example, when `Metric Enum` is 0(PDUs Received), then `Type/Average Enum` has to be 0(Count/summation). Otherwise it's meaningless(we won't calculate count with exponential moving average). And vice versa for other `Metric Enum`. 

So the current fact is that we can always infer `Type/Average Enum` from `Metric Enum`. `Type/Average Enum` isn't quite useful for now. We can add some check to see if the two fields match. Perhaps in extensions in the future, the two fields could work in a combination way. I think for now the usage of `Type/Average Enum` is just a information for the user about how the metrics is calculated(though it's already implied by `Metric Enum`).

Reply:
Would you please help to share your inputs? @jwhui @edmont

Reply:
@librasungirl maybe this note in _4.11.3.4.4.4 Forward Probing Registration Sub-TLV_ is the answer you are looking for?
> Type/Average Enumeration value='0' MUST be used only in conjunction with Metric Enumeration Value='0'.

Reply:
@edmont , `Type/Average 0` seems always bundle with `Metric Enum 0`, in both Single Probe and Forward Probe.
and `Type /Average 1` always bundles with the other Metric Enum [1, 2, 3]. 
Like @abtink mentioned, there is no significant combination usage given existing metric enums defined in spec.  

Not related to this PR, just from spec's perspective, better to keep `Type/Average Enum` bits as reserved as there are no clear usage for now. May continue the discussion within Thread Group without blocking this PR.

Reply:
JIRA [SPEC-956](https://threadgroup.atlassian.net/browse/SPEC-956) to track this question.